### PR TITLE
Adding support for Amazon ECR

### PR DESCRIPTION
**Pull request by leedm777**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
ecr

##### ANSIBLE VERSION
```
ansible 2.3.0 (amazon-ecr a853f25dfd) last updated 2016/12/13 10:43:10 (GMT -500)
  config file = /Users/dlee/prj/ansible/test/integration/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

This patch adds a new module named ecr, which can create, update or
destroy Amazon EC2 Container Registries. It also handles the management
of ECR policies.

Along with this, I've corrected some of the text in the Amazon modules
GUIDELINES.md file, and I've extracted a boto_exception function into
the ec2 module utils.

### CODE REVIEW

**Comment by kkarimi** [comment]
Looks like a nice addition to the AWS modules, I will try to find some time testing it out. @simplesteph any chance we can get some momentum here too?

File: lib/ansible/module_utils/ec2.py [hunk 1]
```diff
@@ -119,6 +119,19 @@ def _boto3_conn(conn_type=None, resource=None, region=None, endpoint=None, **par
 
 boto3_inventory_conn = _boto3_conn
 
+
+def boto_exception(err):
```
Comment (simplesteph) [inline lib/ansible/module_utils/ec2.py]: I believe any changes you're making to ec2 modules should be part of another PR

File: lib/ansible/modules/cloud/amazon/GUIDELINES.md [hunk 2]
```diff
@@ -112,11 +112,11 @@ if my_new_feauture_Parameter_is_set:
 ### Connecting to AWS
 
 To connect to AWS, you should use `get_aws_connection_info` and then
-`connect_to_aws`.
+`boto3_conn` (or `connect_to_aws` for boto).
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/GUIDELINES.md]: same here, part of another PR please

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        elif state == 'absent':
+            result['name'] = name
+            if repo:
+                ecr.delete_repository(name)
+                result['changed'] = True
+
+    except Exception as err:
+        msg = str(err)
+        if isinstance(err, ClientError):
+            msg = boto_exception(err)
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: implement that function in your script or wait for the other PR to get through. (I recommend the former)

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        return dict(repositoryName=name)
+
+    def set_repository_policy(self, name, policy_text):
+        if not self.check_mode:
+            policy = self.ecr.set_repository_policy(repositoryName=name,
+                                                    policyText=policy_text)
+            self.changed = True
+            return policy
+
+        self.skipped = True
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: can you please have an else block statement here?. 
```
if
  return
else 
  return
```

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        if self.get_repository(name) is None:
+            # TODO: if does not exists, fail gracefully
+            raise Error('TODO: handle this case')
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: Can you please implement this?

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+            raise
+
+    def create_repository(self, name):
+        if not self.check_mode:
+            repo = self.ecr.create_repository(repositoryName=name).get(
+                'repository')
+            self.changed = True
+            return repo
+
+        self.skipped = True
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: can you please have an else block statement here?. 
```
if
  return
else 
  return
```

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+            raise Error('TODO: handle this case')
+        return
+
+    def delete_repository(self, name):
+        if not self.check_mode:
+            repo = self.ecr.delete_repository(repositoryName=name)
+            self.changed = True
+            return repo
+
+        repo = self.get_repository(name)
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: if else block please

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+            return repo
+        return None
+
+    def delete_repository_policy(self, name):
+        if not self.check_mode:
+            policy = self.ecr.delete_repository_policy(repositoryName=name)
+            self.changed = True
+            return policy
+
+        policy = self.get_repository_policy(name)
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: if else block please

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+            self.changed = True
+            return policy
+
+        self.skipped = True
+        if self.get_repository(name) is None:
+            # TODO: if does not exists, fail gracefully
+            raise Error('TODO: handle this case')
+        return
+
+    def delete_repository(self, name):
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: there is a force option and a registry id option. Can we include them in the module?
http://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.delete_repository

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+            self.changed = True
+            return repo
+
+        repo = self.get_repository(name)
+        if repo:
+            self.skipped = True
+            return repo
+        return None
+
+    def delete_repository_policy(self, name):
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: can we include registry id?
http://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.delete_repository_policy

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+            res = self.ecr.describe_repositories(repositoryNames=[name])
+            repos = res.get('repositories')
+            return repos and repos[0]
+        except ClientError as err:
+            code = err.response['Error'].get('Code', 'Unknown')
+            if code == 'RepositoryNotFoundException':
+                return None
+            raise
+
+    def get_repository_policy(self, name):
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: can we include registryId ?
http://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.get_repository_policy

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+        if not self.check_mode:
+            repo = self.ecr.create_repository(repositoryName=name).get(
+                'repository')
+            self.changed = True
+            return repo
+
+        self.skipped = True
+        return dict(repositoryName=name)
+
+    def set_repository_policy(self, name, policy_text):
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: Can we include the missing params registryId and force?
http://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.set_repository_policy

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+            get_aws_connection_info(module, boto3=True)
+
+        self.ecr = boto3_conn(module, conn_type='client',
+                              resource='ecr', region=region,
+                              endpoint=ec2_url, **aws_connect_kwargs)
+        self.check_mode = module.check_mode
+        self.changed = False
+        self.skipped = False
+
+    def get_repository(self, name):
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: can we include registryID please?
http://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.describe_repositories

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+options:
+    name:
+        description:
+            - the name of the repository
+        required: true
+    policy:
+        description:
+            - dict that gets JSONified for policy_text
+        required: false
+    policy_text:
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: I understand we may want `policy` (to provide an actual dict) OR `policy_text` to provide a long string. Have you seen any module that has the same behaviour?

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+        try:
+            if policy_text is not None:
+                # policy_text given; parse it
+                policy = json.loads(policy_text)
+            elif policy is not None:
+                # policy given; dump it
+                policy_text = json.dumps(policy)
+        except ValueError:
+            result['msg'] = 'Invalid JSON given for policy_text'
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: if json.dumps(policy) will never error out, can you please do the following:
```
if policy_text is not None:
   try:
      <>
   except:
elif policy is not None:
   <>

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+
+            elif policy is not None:
+                try:
+                    if verbosity >= 2:
+                        result['policy'] = policy
+                    original_policy = ecr.get_repository_policy(name)
+                    if verbosity >= 3:
+                        result['original_policy'] = original_policy
+
+                    if original_policy != policy:
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: for dictionary equality, are we sure this is going to work?
There is an array within the ECR policy (the statements), I'm not sure how AWS puts them in order? also unsure if AWS orders the keys of the dictionary on write. Maybe it's just fine, need the confirmation

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+            - the name of the repository
+        required: true
+    policy:
+        description:
+            - dict that gets JSONified for policy_text
+        required: false
+    policy_text:
+        description:
+            - string for new policy text
+            - setting to an empty string removes the repository's policy
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: if that's the case, is there a need for `delete_policy`?

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+            - create or destroy the repository
+        required: false
+        choices: [present, absent]
+        default: 'present'
+author:
+ - David M. Lee (@leedm777)
+extends_documentation_fragment: aws
+'''
+
+EXAMPLES = '''
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: Do you mind adding examples for all the use cases?

i.e. it's missing the delete policy one

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+            AWS: '{{ read_only_arn }}'
+          Action:
+            - ecr:GetDownloadUrlForLayer
+            - ecr:BatchGetImage
+            - ecr:BatchCheckLayerAvailability
+
+- name: set-policy-text
+  ecr:
+    name: needs-policy
+    policy_text: "{{ lookup('template', 'policy.json.j2') }}"
```
Comment (simplesteph) [inline lib/ansible/modules/cloud/amazon/ecr.py]: the thing is, if we remove the policy_text option, we can still do the following?

```
policy: "{{ lookup('template', 'policy.json.j2') | to_json }}"
```

**Review by simplesteph — CHANGES_REQUESTED** [review]
Overall, really good work. I have done quite a few comments but the goal is to get something that shouldn't require PRs until AWS change their API. 

My feedback:
- Separate in two PR if you're changing other modules
- rename `ecr.py` to `ecs_ecr.py` for consistency and because ecr is still under ecs in the aws console. 
- a few coding style issues for readability and a few questions, please address those. 

I haven't tested it yet, but it looks good otherwise. Thanks for the tests. Implement the changes, then I'll test the module on my own playbooks, and if all good I'll approve. 

Let me know your thoughts,
Stephane

**Comment by ryansb** [comment]
It looks like #19273 is adding the same module -- can community members provide any feedback on which approach is preferred?

**Comment by simplesteph** [comment]
Hi @ryansb I'd favour this PR over #19273 as it has support for policy documents and better matches the API from boto (as in we'll have less PR following this one because it's more complete). I'd like @leedm777 to incorporate my feedback and see this merged ASAP :)

**Comment by leedm777** [comment]
Thanks for the feedback! I've been out/recovering b/c of the holidays. I'll address feedback and update the PR soon-ish.

File: lib/ansible/module_utils/ec2.py (hunk 1 above)
Comment (simplesteph) [reply lib/ansible/module_utils/ec2.py]: so actually I misread the file name (didn't see it was in modules_utils), this would be the right place for such a function. Unsure if other modules would benefit from it, I don't mind it there or in your ecr code.

File: lib/ansible/modules/cloud/amazon/GUIDELINES.md (hunk 2 above)
Comment (leedm777) [reply lib/ansible/modules/cloud/amazon/GUIDELINES.md]: Done. See #20305

File: lib/ansible/modules/cloud/amazon/ecr.py (hunk 14 above)
Comment (leedm777) [reply lib/ansible/modules/cloud/amazon/ecr.py]: No, but I've certainly wished for it. Especially with issues like ansible/ansible-modules-extras#1813, which make it difficult to put JSON in a template.

File: lib/ansible/modules/cloud/amazon/ecr.py (hunk 17 above)
Comment (leedm777) [reply lib/ansible/modules/cloud/amazon/ecr.py]: Sorry; that's an old comment. That was my original intention, but I didn't like the final result so went with `delete_policy` instead. Will remove.

File: lib/ansible/modules/cloud/amazon/ecr.py (hunk 19 above)
Comment (leedm777) [reply lib/ansible/modules/cloud/amazon/ecr.py]: Agreed. I can remove `policy_text`, which can be easily added back if you find a need for it.

File: lib/ansible/modules/cloud/amazon/ecr.py [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+    return True, result
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        name=dict(required=True),
+        state=dict(required=False, choices=['present', 'absent'],
+                   default='present'),
+        policy=dict(required=False, type='dict'),
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ecr.py]: This is a great use case for the `json` type that Ansible offers, so you can either pass valid JSON (such as from a file, or any other source) or a YAML dictionary defined in the task/play and Ansible will handle converting to a Python dict, eliminating the need for having both `policy` and `policy_text` options

File: lib/ansible/modules/cloud/amazon/ecr.py (hunk 19 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ecr.py]: Please have a look at my comment later on about the `json` data type.

File: lib/ansible/modules/cloud/amazon/ecr.py (hunk 14 above)
Comment (simplesteph) [reply lib/ansible/modules/cloud/amazon/ecr.py]: see comment from @ryansb

File: lib/ansible/modules/cloud/amazon/ecr.py (hunk 16 above)
Comment (leedm777) [reply lib/ansible/modules/cloud/amazon/ecr.py]: Good catch. Looks like the guidelines have been updated to recommend `sort_json_policy_dict()`, which should be sufficient for comparing policy docs.

**Comment by leedm777** [comment]
@ryansb @simplesteph Thanks for the feedback. I've just pushed a commit that addresses everything. Let me know if there's anything else to take care of.

File: lib/ansible/modules/cloud/amazon/ecs_ecr.py [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        name=dict(required=True),
+        registry_id=dict(required=False),
+        state=dict(required=False, choices=['present', 'absent'],
+                   default='present'),
+        force_set_policy=dict(required=False, type='bool', default=False),
+        policy=dict(required=False, type='json'),
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ecs_ecr.py:356]: :+1:

File: lib/ansible/modules/cloud/amazon/ecs_ecr.py [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+'''
+
+RETURN = '''
+state:
+    type: string
+    description: The asserted state of the repository (present, absent)
+created:
+    type: boolean
+    description: If true, the repository was created
+name:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ecs_ecr.py:126]: Is there a reason name isn't always included?

**Comment by ryansb** [comment]
Looks great, thank you for contributing!
